Source/AwsCommonRuntimeKit/mqtt/Mqtt5Client.swift

```diff
@@ -259,43 +313,56 @@ public class Mqtt5Client {
                 let continuationCore = ContinuationCore(continuation: continuation)
                 callbackOptions.completion_callback = unsubscribeCompletionCallback
                 callbackOptions.completion_user_data = continuationCore.passRetained()
-                let result = aws_mqtt5_client_unsubscribe(rawValue, unsubscribePacketPointer, &callbackOptions)
-                guard result == AWS_OP_SUCCESS else {
-                    continuationCore.release()
-                    return continuation.resume(throwing: CommonRunTimeError.crtError(CRTError.makeFromLastError()))
+                self.rwlock.read {
+                    // Validate close() has not been called on client.
+                    guard let rawValue = self.rawValue else {
+                        continuationCore.release()
+                        return continuation.resume(throwing: CommonRunTimeError.crtError(CRTError.makeFromLastError()))
+                    }
+                    let result = aws_mqtt5_client_unsubscribe(rawValue, unsubscribePacketPointer, &callbackOptions)
+                    guard result == AWS_OP_SUCCESS else {
+                        continuationCore.release()
+                        return continuation.resume(throwing: CommonRunTimeError.crtError(CRTError.makeFromLastError()))
+                    }
                 }
             }
         }
     }
 
+    /// Discard all operations and cleanup the client. It is MANDATORY function to call to release the client.
     public func close() {
```

Comment:
In other words, making it mandatory to call close just pushes the problem to customers to figure out when it is safe to deinit the client. We should try to solve this problem internally because it's a hard problem. I imagine something along the following lines.
MQTTClient has an external ref and an internal ref.

1. External ref goes to zero with ARC, the MQTT client is in the closing state and the customer cannot do anything new with it like publish, etc.
2. All the pending stuff will either complete successfully or with an error (if the client was stopped, etc.)
3. Internal ref goes to zero, and the client is cleaned up.

Reply:
The internal/external reference approach does not serve as a substitute for the close() function in this context. Instead, the close() function targets to resolve the issue where the external reference count never reaches zero.

If the client is offline or has any network issue, the operations would never completes and be left in the queue. As the operations use continuation to follow the async API design, consequentially, the ARC never goes to 0 in such case (external ref never goes to 0). The close() is used to release the external ref in such case. 
And it is not something we can decide internally, as the user might choose to wait forever on the server responds or they could timeout in certain case. They should be able to decide which way to go.

Another possible approach is that we can provide a Client Timeout concept which will release the client after a certain time. However, I think it will cause confusion between mqtt actual operation timeout (which failed the operation after certain time), and the binding client timeout (which release the client if the operation not completes). I don't feel like to introduce a new concept here.

Reply:
> The internal/external reference approach does not serve as a substitute for the close() function in this context. Instead, the close() function aims to resolve the issue where the external reference count never reaches zero.

@xiazhvera Yeah, I agree and am not saying that we should get rid of close. It should be an optional function, which means that it cancels anything pending, etc. It should not always be required. If the customer lets go of the last reference, it should still clean up everything instead of leaking memory. I can think of the following choices when the last reference is dropped:
1. Make it call the close function, i.e., cancel everything pending.
2. Wait for the operations to complete (maybe with a timeout value?) and clean up everything after the timer is up or all the pending operations are completed.

> If the client is offline or has any network issues, the operations would never complete and be left in the queue. As the operations use continuation to follow the async API design, consequently, the ARC never goes to 0 in such cases (external reference never goes to 0).

The problem is by introducing a mandatory close function, we are just pushing the problem to the customer. Should they keep waiting for the pending operations or timeout, etc.? Instead of us handling this problem for them, every customer will have to develop a way to keep track of the pending operations, implement a timeout function, and then close the client after the timeout.

> Another possible approach is that we can provide a Client Timeout concept, which will release the client after a certain time. However, I think it will cause confusion between the MQTT actual operation timeout (which fails the operation after a certain time) and the binding client timeout (which releases the client if the operation does not complete). I don't feel like introducing a new concept here.

Honestly, I don't mind this. If the MQTT actual operation has a default timeout, we don't even need to introduce a client timeout. The client will get cleaned up when the MQTT operation times out. If there is no default timeout, we can introduce one or just document that customers should set it if they don't want to wait forever.

Reply:
>  the ARC never goes to 0 in such cases (external reference never goes to 0).
The ARC should go to zero unless the customer is keeping an explicit reference to the client or capturing it in the callback, etc. (which is a tricky case). From our side, pending callbacks should not acquire a reference to the client or an external ref if we are using internal/external reference.